### DISCUSSION TITLE
[Linux] Abnormally low frame rate on non-experimental OpenGL, experimental Vulkan won't start

### DISCUSSION THREAD

**Discussion by M1N0RM1N3R** [ppy/osu#28044] [Q&A] [answered]
After finding out that it runs natively on Linux, I finally took the plunge into osu!
It is a fun game, but I have one issue that will eventually become a full-blown problem down the line:
My game runs at a much slower frame rate than I would expect given my hardware setup and performance in other games, including some dips that seriously throw me off when there's more than a couple of notes on screen at a time. (though that could just be a skill issue--I literally got this game less than 24 hours ago :sweat_smile:  )
The experimental Vulkan renderer also doesn't seem to jive with my GPU--it reverts back to OpenGL as soon as I restart.

To help with troubleshooting, I have gathered some info about my (current, subject-to-change) setup:
<details><summary>Steam System Information</summary>
<p>

Computer Information:
Manufacturer: GIGABYTE
Model: AORUS 15G WB
Form Factor: Laptop
No Touch Input Detected
Processor Information:
CPU Vendor: GenuineIntel
CPU Brand: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
CPU Family: 0x6
CPU Model: 0xa5
CPU Stepping: 0x2
CPU Type: 0x0
Speed: 5000 MHz
12 logical processors
6 physical processors
Hyper-threading: Supported
FCMOV: Supported
SSE2: Supported
SSE3: Supported
SSSE3: Supported
SSE4a: Unsupported
SSE41: Supported
SSE42: Supported
AES: Supported
AVX: Supported
AVX2: Supported
AVX512F: Unsupported
AVX512PF: Unsupported
AVX512ER: Unsupported
AVX512CD: Unsupported
AVX512VNNI: Unsupported
SHA: Unsupported
CMPXCHG16B: Supported
LAHF/SAHF: Supported
PrefetchW: Unsupported
Operating System Version:
Pop!_OS 22.04 LTS (64 bit)
Kernel Name: Linux
Kernel Version: 6.8.0-76060800daily20240311-generic
X Server Vendor: The X.Org Foundation
X Server Release: 12101004
X Window Manager: GNOME Shell
Steam Runtime Version: steam-runtime_0.20240304.79797
Video Card:
Driver: NVIDIA Corporation NVIDIA GeForce RTX 2070 with Max-Q Design/PCIe/SSE2
Driver Version: 4.6.0 NVIDIA 550.67
OpenGL Version: 4.6
Desktop Color Depth: 24 bits per pixel
Monitor Refresh Rate: 240 Hz
VendorID: 0x10de
DeviceID: 0x1f14
Revision Not Detected
Number of Monitors: 1
Number of Logical Video Cards: 2
Primary Display Resolution: 1920 x 1080
Desktop Resolution: 1920 x 1080
Primary Display Size: 13.54" x 7.64" (15.51" diag), 34.4cm x 19.4cm (39.4cm diag)
Primary VRAM: 8192 MB
Sound card:
Audio device: Realtek ALC255
Memory:
RAM: 15862 Mb
VR Hardware:
VR Headset: None detected
Miscellaneous:
UI Language: English
LANG: en_US.UTF-8
Total Hard Disk Space Available: 470530 MB
Largest Free Hard Disk Block: 14484 MB
Storage:
Number of SSDs: 3
SSD sizes: 1000G,512G,0B
Number of HDDs: 0
Number of removable drives: 0

</p>
</details> 

[1714505775.performance.log](https://github.com/ppy/osu/files/15171005/1714505775.performance.log)

If you need any more info, please let me know, and I'll update this post accordingly.

E1: Added another newline between HW info and perf log link so you can actually click it...I hear you're a fan of clicking things.
E2: The experimental OpenGL renderer seems to work better--I'm getting triple-digit frame rates fairly consistently, and it rarely if ever went below 60FPS during a test play--though I was focusing on the game itself than the little frame rate counter in the corner, it certainly *felt* smoother.

**Comment by M1N0RM1N3R** [comment]
Just thought I'd give an update (and bump):
Experimental OpenGL still seems to run the best out of the 3 options available to me (one of which *starting to chuckle* doesn't even work at all).
But there is still a difference between "best" and "good"--in my case, quite a significant one, and not the kind with overlap, either.
I quickly screen recorded a playthrough of *Triangles / cYsmix* on *Tt's Normal* map (jc it's the first one I really ran into issues with), and, yyyeah, the frame rate blows.
Edit 1: The video upload has been stuck on "Processing will begin shortly" for HOURS. Deleting and re-uploading doesn't seem to be doing anything. I'll update this post if/when I can get YT to ingest my footage, but for now, take my word for it: it gets BAD.
Edit 2: It took way too long to get my test recording actually accessible, but [here it is](https://www.youtube.com/watch?v=r9WyoN1IxDg).

**Comment by sineplusx** [comment]
Can you check if the game is using the dedicated or the integrated GPU? This should be visible at the top of each `runtime.log`, which you didn't attach here.

Vulkan not starting on Linux is a known issue that will be fixed in the next release. If you still want to try it in the meantime, use the workaround from [this thread](https://github.com/ppy/osu/discussions/27659#discussioncomment-9101487). Vulkan is also a bit unstable currently, you might run into freezes.

**Reply by M1N0RM1N3R** [reply]
A-ha!
![image](https://github.com/ppy/osu/assets/74984624/17ae69cf-6b14-451b-955c-95b0e7c5f54c)
Looks like it's using the iGPU instead of the dGPU.
Now to figure out how to get GNOME (and maybe KDE later) to run an AppImage on the dGPU...

**Reply by M1N0RM1N3R** [reply, marked as answer]
OK, after quite a bit of Googling, I cobbled together a shell script to launch osu! using the dGPU.
By just setting a handful of environment variables, I could switch osu! to run on the dGPU and get the consistent triple-digit frame rates I came to expect from my hardware.
To those of you in a similar boat to me, drop this script into the same directory as your copy of osu! and use it in your .desktop shortcut. It should improve your performance *significantly*.

[Here's my new launcher shell script](https://gist.github.com/M1N0RM1N3R/b5245ff2d0c1206308cd8d963733ed24).
And a screenshot of my runtime.log after launching with this script:
![Screenshot from 2024-05-03 08-35-07](https://github.com/ppy/osu/assets/74984624/b2d84d1e-7a13-4b99-8078-a37ccba6fbca)

You bet your :bacon: this is going on the forum.
